pkg/fqdn/service/service.go

```diff
@@ -102,21 +139,96 @@ type PolicyUpdater interface {
 
 // StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
 // SDP calls this method to subscribe to DNS policies
-// For each stream, we start a goroutine to receive the DNS policies ACKs
+// For each stream, we subscribe to the changes in the policy rules table and send the current state of the DNS rules to the client.
 // The flow of the method is as follows:
-// 1. Add the stream to the map( called by the client i.e SDP)
-// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
-// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
-// 4. Wait for the context to be done
-// Note: this method is left empty on purpose and will be update with the actual implementation in the future PRs for the standalone DNS proxy
+//  1. Send the current state of the DNS rules to the client.
+//  2. Subscribe to the changes in the policy rules table and wait for changes.
+//  3. For each change in the policy rules table, send the current state of the DNS rules to the client.
+//  4. If the stream context is done, return.
 func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
-	// This is a temporary implementation to send the current state of the DNS rules to the client and used for testing
-	stream.Send(&pb.PolicyState{RequestId: "test"})
+	streamCtx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+
+	wtxn := s.db.WriteTxn(s.policyRulesTable)
+	changeIterator, err := s.policyRulesTable.Changes(wtxn)
+	if err != nil {
+		wtxn.Abort()
+		return fmt.Errorf("failed to watch policy rules table: %w", err)
+	}
+	wtxn.Commit()
+
+	// Send the current state of the DNS policies to the client
+	if err := s.sendAndRecvAckForDNSPolicies(stream); err != nil {
+		s.log.Error("Failed to send current state of DNS policies to the client", logfields.Error, err)
+		return err
+	}
+
+	for {
+		// Wait for changes in the policy rules table.
+		// If there are changes, it will return the changes and a channel to watch for new changes.
+		// We will then send the current state of the DNS policies to the client.
+		changes, watch := changeIterator.Next(s.db.ReadTxn())
+		for range changes {
+			err := s.sendAndRecvAckForDNSPolicies(stream)
```

Comment:
Why are we not looking at the changes here? 
Is the idea to change incremental changes or a full snapshot? If full snapshot then we should not be using `Changes()` but instead just do `AllWatch` instead. Rate limiting should also be added to avoid sending intermediate states.